### PR TITLE
ENH: Test for sitkUtils to make sure to push/pull data from Slicer 

### PR DESCRIPTION
**Pull request by MayeulChassagnard**
According to the request [ #1693](http://na-mic.org/Mantis/view.php?id=1693) on MantisBugTracker, I added a test for Push/Pull data from Slicer

A minor error linked with the call of PullFromSlicer (a function from sitkUtils) still exist but do not interfere with the test.

The detail of the error can be seen in the file attached [here](https://github.com/Slicer/Slicer/files/364637/error_PullFromSlicer.zip). Otherwise, when running the following terminal command : `ctest -V -R py_nomainwindow_test_sitkUtils`  in the directory `Slicer-SuperBuild/Slicer-build`

### CODE REVIEW

File: Base/Python/tests/test_sitkUtils.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import slicer
+#import SimpleITK as sitk
+import sitkUtils as su
+
+import unittest
+
+class SitkUtilsTests(unittest.TestCase):
+
+    def runTest(self):
```
Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: This function should not be needed here

File: Base/Python/tests/test_sitkUtils.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        self.assertEqual(self.myNode1.GetMTime(), myNode2.GetMTime(),
+                         'Error Push Pull: Modify Time are not the same')
+
+        """ To perform the test:
+        Try with a modified sitkimage and verify the GetMTime are diffrents
+        Try with all the compositeView and overwrite (all possiblities)
+        """
+
+    def tearDown(self):
+        del self.myNode1
```
Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: I suggest to call `slicer.mrmlScene.Close()` here

File: Base/Python/tests/test_sitkUtils.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import slicer
+#import SimpleITK as sitk
```
Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: Remove commented code

File: Base/Python/tests/test_sitkUtils.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    def setUp(self):
+        """ Download the MRHead node
+        """
+
+        from SampleData import SampleDataLogic
+        SampleDataLogic().downloadMRHead()
+        myNode1 = slicer.util.getNode('MRHead')
+
+        print("Node Name is %s" % myNode1.GetName())
+        self.assertIsNotNone(myNode1.GetName(),"ERROR: Node Name is None !")
```
Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: Instead uses:
```
self.assertEqual(myNode1.GetName(), "MRHead")
```

File: Base/Python/tests/test_sitkUtils.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    def setUp(self):
+        """ Download the MRHead node
+        """
+
+        from SampleData import SampleDataLogic
+        SampleDataLogic().downloadMRHead()
+        myNode1 = slicer.util.getNode('MRHead')
+
+        print("Node Name is %s" % myNode1.GetName())
```
Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: Print statement not needed

File: Base/Python/tests/test_sitkUtils.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        self.test_SimpleITK_SlicerPushPull()
+        self.tearDown()
+
+    def setUp(self):
+        """ Download the MRHead node
+        """
+
+        from SampleData import SampleDataLogic
+        SampleDataLogic().downloadMRHead()
+        myNode1 = slicer.util.getNode('MRHead')
```
Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: `myNode1` -> `volumeNode`

**Comment by jcfr** [comment]
Nice.  Thanks for the PR :+1: 

Look like the error message happen because of the hack done [here](https://github.com/Slicer/Slicer/blob/b09de614b5ebc9e6a8e7974113fe82c637c79a24/Base/Python/sitkUtils.py#L83-L90)

A nice improvement would be to have a more obvious way of registering the IO factories.

@blowekamp What do you think ?

File: Applications/SlicerApp/Testing/Python/CMakeLists.txt [hunk 7]
```diff
@@ -224,6 +224,14 @@ set_tests_properties(py_nomainwindow_${testname} PROPERTIES
   WILL_FAIL TRUE
   )
 
+if(Slicer_USE_SimpleITK)
+slicer_add_python_unittest(
```
Comment (vovythevov) [inline Applications/SlicerApp/Testing/Python/CMakeLists.txt]: This method should be indented (as it's inside a if()/endif())

Comment (MayeulChassagnard) [inline Applications/SlicerApp/Testing/Python/CMakeLists.txt]: Thanks for the review folks !

File: Base/Python/tests/test_sitkUtils.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        self.assertIsNotNone(myNode1.GetName(),"ERROR: Node Name is None !")
+
+        self.myNode1 = myNode1
+
+    def test_SimpleITK_SlicerPushPull(self):
+        """ Try to pull a slicer MRML image, return the SimpleITK,
+        push it to Slicer, and then compare if the image are the same
+        """
+
+        sitkimage = su.PullFromSlicer(self.myNode1.GetName())
```
Comment (blowekamp) [inline Base/Python/tests/test_sitkUtils.py]: I would add some sanity checks for the SimpleITK image, size, pixel type, direction matrix, and at least a couple pixels.

**Comment by blowekamp** [comment]
Perhaps we should ensure EnsureRegistration only gets called once in sitkUtils.py? Or there should be a wrapped C++ method which checks if the registration needs to occur?

File: Base/Python/tests/test_sitkUtils.py (hunk 1 above)
Comment (MayeulChassagnard) [inline Base/Python/tests/test_sitkUtils.py]: So, I can erase runTest ? What about setUp and tearDown ? Must I delete them also ?

**Comment by jcfr** [comment]
> Perhaps we should ensure EnsureRegistration only gets called once in sitkUtils.py?

Calling it once by capturing its stderr would indeed be an improvement

> Or there should be a wrapped C++ method which checks if the registration needs to occur?

Agreed. That would be better

File: Base/Python/tests/test_sitkUtils.py (hunk 1 above)
Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: You need `setUp` and `tearDown`, they are called before and after running each test function.

File: Base/Python/tests/test_sitkUtils.py (hunk 2 above)
Comment (MayeulChassagnard) [inline Base/Python/tests/test_sitkUtils.py]: This command doesn't exist. Do you mean `slicer.mrmlScene.Clear()` instead ?

Comment (jcfr) [inline Base/Python/tests/test_sitkUtils.py]: Ooops. Yess, `slicer.mrmlScene.Clear(0)` it is.

**Comment by jcfr** [comment]
Could you rebase  and push force your topic. This would ensure the CI will success.

```
git svn rebase
git push origin name_of_topic --force
```

**Comment by MayeulChassagnard** [comment]
rebase + push done

**Comment by jcfr** [comment]
Closing. Tests have been added by @lassoan  in [r26230](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=26230)  (f66714c)

**Pull request closed without merging**